File: python/dcbor/src/dcbor/string_util.py
from __future__ import annotations

import unicodedata


def is_nfc(s: str) -> bool:
    return unicodedata.is_normalized("NFC", s)


def to_nfc(s: str) -> str:
    return unicodedata.normalize("NFC", s)


def flanked(s: str, left: str, right: str) -> str:
    return left + s + right


def _is_printable(c: str) -> bool:
    code = ord(c)
    return code > 127 or 32 <= code <= 126


def sanitized(s: str) -> str | None:
    has_printable = False
    chars: list[str] = []
    for c in s:
        if _is_printable(c):
            has_printable = True
            chars.append(c)
        else:
            chars.append(".")
    if not has_printable:
        return None
    return "".join(chars)

File: csharp/DCbor/DCbor/StringUtil.cs
namespace BlockchainCommons.DCbor;

/// <summary>
/// Internal string utility functions for diagnostic formatting.
/// </summary>
internal static class StringUtil
{
    internal static string Flanked(string s, string left, string right)
    {
        return left + s + right;
    }

    internal static bool IsPrintable(char c)
    {
        return !char.IsAscii(c) || (c >= 32 && c <= 126);
    }

    internal static string? Sanitized(string s)
    {
        bool hasPrintable = false;
        var chars = new char[s.Length];
        for (int i = 0; i < s.Length; i++)
        {
            if (IsPrintable(s[i]))
            {
                hasPrintable = true;
                chars[i] = s[i];
            }
            else
            {
                chars[i] = '.';
            }
        }
        return hasPrintable ? new string(chars) : null;
    }
}
